substrate/validator-sets/pallet/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+    // Limit each set to 100 participants for now
+    participants: BoundedVec<(T::AccountId, Amount), ConstU32<100>>,
+  }
+
+  #[pallet::pallet]
+  #[pallet::generate_store(pub(super) trait Store)]
+  pub struct Pallet<T>(PhantomData<T>);
+
+  /// The details of a validator set instance.
+  #[pallet::storage]
+  #[pallet::getter(fn validator_set)]
+  pub type ValidatorSets<T: Config> =
+    StorageMap<_, Twox64Concat, ValidatorSetInstance, ValidatorSet<T>, OptionQuery>;
+
+  type Key = BoundedVec<u8, MaxKeyLen>;
+
+  /// The key for a given validator set instance coin.
+  #[pallet::storage]
+  #[pallet::getter(fn key)]
+  pub type Keys<T: Config> =
+    StorageMap<_, Twox64Concat, (ValidatorSetInstance, Coin), Key, OptionQuery>;
+
+  /// If an account has voted for a specific key or not. Prevents them from voting multiple times.
+  #[pallet::storage]
+  #[pallet::getter(fn voted)]
+  pub type Voted<T: Config> = StorageMap<_, Blake2_128Concat, (T::AccountId, Key), (), OptionQuery>;
+
+  /// How many times a key has been voted for. Once consensus is reached, the keys will be adopted.
+  #[pallet::storage]
+  #[pallet::getter(fn vote_count)]
+  pub type VoteCount<T: Config> =
+    StorageMap<_, Blake2_128Concat, (ValidatorSetInstance, Coin, Key), u16, ValueQuery>;
+
+  #[pallet::genesis_build]
+  impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+    fn build(&self) {
+      // Start at 1 to leave 0 available for a future design where Serai itself has a distinct
+      // validator set for its Tendermint consensus
+      let mut coins = Vec::new();
+      for coin in 0 .. self.coins.0 {
+        coins.push(Coin(coin));
+      }
+
+      let mut participants = Vec::new();
+      for participant in self.participants.clone() {
+        participants.push((participant, self.bond));
+      }
+
+      ValidatorSets::<T>::set(
+        ValidatorSetInstance(Session(0), ValidatorSetIndex(1)),
+        Some(ValidatorSet {
+          bond: self.bond,
+          coins: BoundedVec::try_from(coins).unwrap(),
+          participants: BoundedVec::try_from(participants).unwrap(),
+        }),
+      );
+    }
+  }
+
+  #[pallet::event]
+  #[pallet::generate_deposit(pub(super) fn deposit_event)]
+  pub enum Event<T: Config> {
+    Vote {
+      voter: T::AccountId,
+      instance: ValidatorSetInstance,
+      coin: Coin,
+      key: Key,
+      // Amount of votes the key now has
+      votes: u16,
+    },
+    KeyGen {
+      instance: ValidatorSetInstance,
+      coin: Coin,
+      key: Key,
+    },
+  }
+
+  #[pallet::error]
+  pub enum Error<T> {
+    /// Validator Set doesn't exist.
+    NonExistentValidatorSet,
+    /// Non-validator is voting.
+    NotValidator,
+    /// Validator Set already generated keys.
+    AlreadyGeneratedKeys,
+    /// Vvalidator has already voted for these keys.
+    AlreadyVoted,
+  }
+
+  #[pallet::call]
+  impl<T: Config> Pallet<T> {
+    #[pallet::weight(0)] // TODO
+    pub fn vote(
```

Comment:
would it be logical to separate keygen vote and churn vote? I understand we are targeting single set, single generation for now, but this function seems incompatible with churning.

Reply:
What "churn" vote?

vote isn't incompatible with multiple sets since its bound to a set instance. When a new instance confirms their key, it'll trigger a churn from the old multisig to the new one.

Reply:
Understood, didn't know if we needed a distinction between the genesis keygen of a set versus churning keygens.